### PR TITLE
Implement native stack unwinding for Linux

### PR DESCRIPTION
**Pull request by janvorli**
This change implements native stack unwinding using the libunwind on
Linux. I have also fixed bunch of issues / details in the related code:
1) 0x in front of %p inside format string
2) Subtraction of -1 from dl_info.dli_sname
3) Added .cfi_xxxx annotation to the CallDescrWorkerInternal and the
LEAF_ENTRY / LEAF_END macros.
4) Changed local labels in the CallDescrWorkerInternal to be prefixed by
.L to see the CallDescrWorkerInternal in the stack trace
5) Changed moveOWord to use movdqu - it was being called with one of the
parameters unaligned

### CODE REVIEW

**Comment by janvorli** [comment]
@jkotas Could you please take a look?

**Comment by janvorli** [comment]
@kangaroo This is what works on Linux with the libunwind from http://www.nongnu.org/libunwind. It seems that what we need to change for OSX is the WinContextToUnwindContext function in seh-unwind.cpp. We should be able to use unw_set_reg on OSX for all registers with the OSX libunwind (provided it works the same way as the libunwind in LLVM), while with the other libunwind, e.g. the RSP is considered read only.

**Comment by janvorli** [comment]
@kangaroo This unix_issue177 branch is the branch that I've created for you and me to work on the unwinder.

**Comment by kangaroo** [comment]
@janvorli Awesome!  I'll try to take a look tonight.  For #5 see: https://github.com/dotnet/coreclr/issues/141

**Comment by kangaroo** [comment]
Also, what (if any) tests are you validating against right now?

**Comment by janvorli** [comment]
@kangaroo No tests yet. I've just created a simple C# app that dumps managed stack trace. Like this (needs to be compiled with optimizations off so that the compiler preserves all the methods and the dump doesn't contain just main):
```c#
    class Program
    {
        static void A()
        {
            B();
        }

        static void B()
        {
            C();
        }

        static void C()
        {
            D();
        }

        static void D()
        {
            Console.WriteLine(System.Environment.StackTrace);
        }

        static void Main(string[] args)
        {
            A();
        }
    }
```

File: src/pal/inc/rt/palrt.h [hunk 1]
```diff
@@ -1705,6 +1705,8 @@ typedef LONG (WINAPI *PTOP_LEVEL_EXCEPTION_FILTER)(
     );
 typedef PTOP_LEVEL_EXCEPTION_FILTER LPTOP_LEVEL_EXCEPTION_FILTER;
 
+BOOL VirtualUnwind(CONTEXT *context, KNONVOLATILE_CONTEXT_POINTERS *contextPointers);
```
Comment (jkotas) [inline src/pal/inc/rt/palrt.h]: Maybe call it PAL_VirtualUnwind to make it clear that it is not a regular Win32 API?

File: src/vm/amd64/calldescrworkeramd64.S [hunk 2]
```diff
@@ -42,8 +42,12 @@ NESTED_END FastCallFinalizeWorker, _TEXT
 
 NESTED_ENTRY CallDescrWorkerInternal, _TEXT, CallDescrWorkerUnwindFrameChainHandler
         push_nonvol_reg rbx             // save nonvolatile registers
+        .cfi_def_cfa_offset 16          //
```
Comment (jkotas) [inline src/vm/amd64/calldescrworkeramd64.S]: The .cfi annotations should be in the prolog setup/teardown macros. Something like:

.macro push_nonvol_reg Register
        push \Register
        .cfi_adjust_cfa_offset 8
        .cfi_offset \Register, 0
.endm

Similar for others. Ideally, we should not have any direct use of .cfi annotations - we may need to introduce a few macros to make it work (e.g. pop_nonvol_reg).

Comment (jkotas) [inline src/vm/amd64/calldescrworkeramd64.S]: LGTM, modulo comments.

**Comment by kangaroo** [comment]
I'll try to address my review comments with a patch after dinner.

**Comment by kangaroo** [comment]
Initial PR to clean up OS X here: https://github.com/dotnet/coreclr/pull/284

It still needs a bit of work, since its not actually displaying the stack trace yet, but it doesn't break build or crash anymore.

File: src/pal/inc/rt/palrt.h (hunk 1 above)
Comment (janvorli) [inline src/pal/inc/rt/palrt.h]: Makes sense

**Comment by janvorli** [comment]
@kangaroo 
>This API also isn't available on mac. From the lower context it looks like we're trying to see a register spill to a stack slot, and ensure we load from there instead.

>Shouldn't unw_get_reg take care of this internally?

GC currently relies on having these context pointers so that when it moves an object to which a register stored somewhere on the stack points, it can update the value of the register on the stack. So what we are trying to do there is to get a pointer to the stack location that we can keep and GC can update it later.
The fact that we cannot get the pointers using the API provided by the libunwind on OSX will require some likely nontrivial changes in the coreclr. We have already discussed it a bit with @jkotas some time ago when I was looking for the libunwind to use on Linux. 
One option we have here is to ensure that object pointed to by registers are always pinned. Then the context pointer are unnecessary. It is not clear to me though what exactly needs to be done to do that.
Now that I am thinking about it, I actually wonder if we really need the context pointers for the native frames. @jkotas, do you know if there is a case where a pointer to an object managed by GC would be stored in a register stored in one of the native frames?

**Comment by kangaroo** [comment]
@dotnet-bot test this please

**Comment by jkotas** [comment]
> do you know if there is a case where a pointer to an object managed by GC would be stored in a register stored in one of the native frames

Yes, pointer to an object managed by GC would be stored in a register stored in one of the native frames.

> One option we have here is to ensure that object pointed to by registers are always pinned

If the unwind on Mac does not support context pointers, I would not worry about it right now. It will  become a problem only once we start enabling GC - I would do it as part of enabling GC.

>  It is not clear to me though what exactly needs to be done to do that.

It would require: keeping track of unavailable register context pointers in REGDISPLAY - it may be as simple as storing NULL as register context pointer when it is not available, and then using that information to pin the values that cannot be updated in EECodeManager::EnumGcRefs.

**Comment by janvorli** [comment]
@jkotas I've reflected your feedback in https://github.com/dotnet/coreclr/commit/cf3213d9f47b59dc553b04b5704acd025add1cdc, could you take a look please?

**Comment by janvorli** [comment]
@kangaroo Our changes are now in the master branch.

**Comment by kangaroo** [comment]
@janvorli excellent, I've been looking at the existing machinery around exception handling expecting thats up next ;)

**Comment by janvorli** [comment]
@kangaroo Regarding the exception handling. I've started to work on a plan for it and I've figured that most of the stuff in PAL like the PAL_TryExcept and all the related stuff will go away. Only the exception raising stuff will stay (RtlpRaiseException, SEHRaiseException, RaiseException). We will use plain C++ exceptions instead of this low level simulation of SEH. So hopefully there should be no platform specific stuff for that.
There will be changes needed in the coreclr itself though.

**Comment by pgavlin** [comment]
@janvorli Do you have a timeline for the PAL changes?

**Comment by kangaroo** [comment]
@janvorli Yes, a lot of the existing stuff looked overly complicated.  I think C++ exceptions are the right way to go.

**Comment by AndyAyersMS** [comment]
@janvorli can you keep us informed as the EH plans evolve?

**Comment by janvorli** [comment]
@pgavlin @AndyAyersMS @kangaroo Let's move further discussion to #192 which @sergiy-k has created for the exception handling work.

**Comment by janvorli** [comment]
@pgavlin I have already made the changes in PAL and the related PAL_TRY, PAL_EXCEPT, ... macros locally and I'd like to send out a PR for these today.

**Comment by janvorli** [comment]
@AndyAyersMS I hope I'll be able to share all the plan details tomorrow.

**Pull request merged**